PR for Module 3, functionality 3.

taidy-marrugo opened the pull request.

taidy-marrugo committed:
create layout for fragment and update dimens  values

taidy-marrugo committed:
create field ID for character

taidy-marrugo committed:
invoke service for detail character and use case for it

taidy-marrugo committed:
mvp for feature

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/mapper/CharacterMapperService.kt:
```diff
@@ -10,13 +10,15 @@ open class CharacterMapperService : BaseMapperRepository<CharacterResponse, Char
             transformToThumbnail(characterResponse.thumbnail)
     )
 
     override fun transformToResponse(type: Character): CharacterResponse
```
Kotlin is an infered type language you could do something like:
```
override fun transformToResponse(type: Character) = CharacterResponse(
            type.id,
            type.name,
            type.description,
            transformToThumbnailResponse(type.thumbnail)
```

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterServicesImpl.kt:
```diff
@@ -13,9 +13,21 @@ class CharacterServicesImpl(private val api: MarvelResquestGenerator = MarvelRes
         return Observable.create { subscriber ->
```
Please don't use create methods. Use fromCallable. Also, you can use Single instead Observable because you will emit only one observable.

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterServicesImpl.kt:
```diff
@@ -13,9 +13,21 @@ class CharacterServicesImpl(private val api: MarvelResquestGenerator = MarvelRes
+    }
 
+    fun getCaractersDetail(id: Int?): Observable<List<Character>> {
+        return Observable.create { subscriber ->
```
Same here, don't use create and observable

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterDetailPresenter.kt:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    private fun requestGetCharacterDetail() {
+        val subscription = getChatacterDetailServiceUseCase.invoke(view.idCharacter).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe({ character ->
```
I think you not need to use the invoke word. Put only:
```
getChatacterDetailServiceUseCase(view.idCharacter)
```

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterDetailView.kt:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    fun showToastNoItemToShow() {
+        val fragment = fragmentRef.get()
```
You can use instead:
```
fragment.get()?.{ fragment ->
```

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterDetailView.kt:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    fun showCharactersDetail(character: Character) {
+        val fragment = fragmentRef.get()
```
You can use instead:
```
fragment.get()?.{ fragment ->
```

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharecterView.kt:
```diff
@@ -22,32 +26,29 @@ class CharecterView(activity: MainActivity) {
     }
 
     fun showToastNoItemToShow() {
         val activity = activityRef.get()
```
You can use instead:
```
activityRef.get()?.{ fragment ->
```

gerardosuarezg commented on app/src/main/res/layout/fragment_dialog_detal_character.xml:
```diff
@@ -0,0 +1,33 @@
+            android:textColor="@android:color/white"
+            android:textSize="@dimen/fragment_dialog_text_size" />
+    </LinearLayout>
+
```
Please remove extra space

gerardosuarezg commented on app/src/main/res/layout/fragment_dialog_detal_character.xml:
```diff
@@ -0,0 +1,33 @@
+            android:textSize="@dimen/fragment_dialog_text_size" />
+
+        <TextView
+            android:id="@+id/description"
```
Use the following convention to name the views:
what_where_description
In this case this would be: text_dialog_detail_character_description

gerardosuarezg commented on app/src/main/res/layout/fragment_dialog_detal_character.xml:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Rename this file to: fragment_dialog_detail_character.

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/domain/usecase/GetCharacterDetailServiceUseCase.kt:
```diff
@@ -0,0 +1,10 @@
+import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
+import io.reactivex.Observable
+
+open class GetCharacterDetailServiceUseCase(private val characterServiceImp: CharacterServicesImpl) {
```
This class must not be open

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/base/BaseRxFragmentDialog.kt:
```diff
@@ -0,0 +1,20 @@
+import android.app.DialogFragment
+import io.reactivex.disposables.CompositeDisposable
+
+open class BaseRxFragmentDialog : DialogFragment() {
```
This class must not be open

taidy-marrugo replied:
BaseRxFragmentDialog  is needed open for the implementation on DetailCharacterFragmentDialog and the subscriptions

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharecterView.kt:
```diff
@@ -7,13 +7,17 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
     private val activityRef = WeakReference(activity)
     private val SPAN_COUNT = 1
-    var adapter = CharacterAdapter { character -> activity.applicationContext.showToast(character.name) }
+    var adapter = CharacterAdapter { character ->
```
What if in the future we need some data to init the adapter? You must initialize this in the init() method

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharecterView.kt:
```diff
@@ -7,13 +7,17 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
 import kotlinx.android.synthetic.main.activity_main.*
 import java.lang.ref.WeakReference
 
 class CharecterView(activity: MainActivity) {
```
Typo. Must be CharacterView instead

taidy-marrugo committed:
changes on res

remove extra space
change file name

taidy-marrugo committed:
changes on logic

-remove invoke  on CharacterDetailPresenter
-rename file  CharacterView
-change initization from adapter on CharacterView

taidy-marrugo committed:
update get

-changes on transformToResponse

taidy-marrugo committed:
refactor fragment and activity instance

taidy-marrugo committed:
update observable to fromCallable

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterServicesImpl.kt:
```diff
@@ -1,25 +1,30 @@
+            return Observable.fromCallable(Callable<List<Character>>() {
+                val callResponse = api.createService(MarvelApi::class.java).getCharacter()
+                val response = callResponse.execute()
+                return@Callable  mapper.transform(response.body()?.data!!.characters)
```
Please don't use !!

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/CharacterServicesImpl.kt:
```diff
@@ -1,25 +1,30 @@
+        return Observable.fromCallable(Callable<List<Character>>() {
+            val callResponse = api.createService(MarvelApi::class.java).getCharacterDetail(id)
+            val response = callResponse.execute()
+            return@Callable  mapper.transform(response.body()?.data!!.characters)
```
Please don't use !!

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/api/MarvelApi.kt:
```diff
@@ -5,9 +5,12 @@ import com.puzzlebench.clean_marvel_kotlin.data.service.response.DataBaseRespons
     @GET("/v1/public/characters")
     fun getCharacter(): Call<MarvelBaseResponse<DataBaseResponse<ArrayList<CharacterResponse>>>>
+    @GET("/v1/public/characters/{id}")
+    fun getCharacterDetail(@Path("id") id: Int?): Call<MarvelBaseResponse<DataBaseResponse<List<CharacterResponse>>>>
```
Is it really possible the parameter being null? I think you must validate the nullability before call this method. Please change Int? by Int

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/fragment/DetailCharacterFragmentDialog.kt:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    companion object {
+        @JvmStatic
```
If the project is full developed with kotlin, why do you need to add this annotation?

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterDetailView.kt:
```diff
@@ -0,0 +1,38 @@
+    var idCharacter: Int? = null
+
+    fun init() {
+        val fragment = fragmentRef.get()
```
Please refactor this using the let function

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterDetailView.kt:
```diff
@@ -0,0 +1,38 @@
+    fun showCharactersDetail(character: Character) {
+            fragmentRef.get()?.text_dialog_detail_character_name?.text = character.name
+            fragmentRef.get()?.text_dialog_detail_character_description?.text = character.description
+            val string = character.thumbnail.path + "." + character.thumbnail.extension
```
Use string templates instead + operators

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,45 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
-
+    private var adapter: CharacterAdapter? = null
     fun init() {
         val activity = activityRef.get()
```
Use let operator to simplify this code

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,45 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
-            activity.applicationContext.showToast(message)
-
-        }
+        val message = activityRef.get()?.baseContext?.resources?.getString(R.string.message_no_items_to_show)
```
You don't need to get the baseContext to get the resources. You can get it from activity.

gerardosuarezg replied:
What if you refactor this code:
```
   activity.resources?.getString(R.string.message_no_items_to_show)?.let{ activity.showToast(it)  }  
```

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,45 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
 
-class CharecterView(activity: MainActivity) {
+class CharacterView(activity: MainActivity) {
     private val activityRef = WeakReference(activity)
```
You don't need to hold a weak reference to the activity. If you subscribe/unsubscribe the presenter from the activity/fragment you can assume the activity never will be null. 
Please mark the activity as private val and use it instead this weak reference in this class. So, you won't need to do this anymore: `activityRef.get()?`

gerardosuarezg commented on app/src/main/res/layout/fragment_dialog_detail_character.xml:
```diff
@@ -0,0 +1,32 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent" />
+
+    <LinearLayout
```
Please don't use a LinearLayout inside a relative layout, this will decrease the performance of the app. If possible, please refactor to constraintLayout. In other case, only use the RelativeLayout and remove the use of LinearLayout

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterDetailView.kt:
```diff
@@ -0,0 +1,38 @@
+import java.lang.ref.WeakReference
+
+class CharacterDetailView(fragment: DetailCharacterFragmentDialog) {
+    private val fragmentRef = WeakReference(fragment)
```
Same here, please remove this weak reference and use the fragment you got as parameter

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterDetailPresenter.kt:
```diff
@@ -0,0 +1,28 @@
+class CharacterDetailPresenter(view: CharacterDetailView, private val getChatacterDetailServiceUseCase: GetCharacterDetailServiceUseCase, val subscriptions: CompositeDisposable) : Presenter<CharacterDetailView>(view) {
+    fun init() {
+        view.init()
+        requestGetCharacterDetail()
```
Please remove the call for the request method from the init and call it from activity/fragment

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterDetailView.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    fun showToastNetworkError(error: String) {
+        fragmentRef.get()?.activity?.showToast(error)
```
Remove weak reference across the class

taidy-marrugo committed:
improves on code

- remove  jvmstatic
-remove + operators for concat and replace for plus
-remove baseContext  from showToastNoItemToShow
-refactor string message on showToastNoItemToShow
-remove  val activityRef = WeakReference(activity) and
activityRef.get()
-remove  val fragmentRef = WeakReference(fragment) and
fragmentRef.get()
-change call requestGetCharacterDetail() from presenter to fragment

taidy-marrugo committed:
layout, callable and type  of variable

-refactor layout
- remove !! from  CharacterServicesImpl
- change Int? to int and validate value

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/data/service/api/MarvelApi.kt:
```diff
@@ -5,9 +5,12 @@ import com.puzzlebench.clean_marvel_kotlin.data.service.response.DataBaseRespons
 import retrofit2.Call
 import retrofit2.http.GET
+import retrofit2.http.Path
 import java.util.*
```
Please don't use wildcard imports

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/adapter/viewholder/viewholder.kt:
```diff
@@ -12,7 +12,7 @@ class CharactersAdapterViewHolder(view: View, val listener: CharacterListener) :
     fun bind(item: Character) = with(itemView) {
         tv_item.text = item.name
-        val string = item.thumbnail.path + "." + item.thumbnail.extension
+        val string =item.thumbnail.path.plus(".").plus(item.thumbnail.extension)
```
Please use String templates to do that. Check this example:
```
fun main(args: Array<String>) {

    val str1 = "hello"
    val str2 = "hi"
    val str3 = "bye"

    // string interpolation
    val str4 = "$str1 $str2 $str3"

    // displaying concatenated string
    println(str4)
}
```

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,44 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
-class CharecterView(activity: MainActivity) {
-    private val activityRef = WeakReference(activity)
+class CharacterView(activity: MainActivity) {
+    private  val activity = activity
```
You don't need this reference to the activity, you have the injected reference. Please remove this variable and refactor the constructor as follows:
```
class CharacterView(private val activity: MainActivity) {
```

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,44 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
     fun init() {
-        val activity = activityRef.get()
+
         if (activity != null) {
```
Please remove the if condition. You don't have to validate if activity is null.

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,44 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
-            activity.applicationContext.showToast(message)
-
-        }
+        activity?.resources?.getString(R.string.message_no_items_to_show)?.let { activity?.applicationContext?.showToast(it) }
```
You don't have to validate if activity is null

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,44 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
 
     fun showToastNetworkError(error: String) {
-        activityRef.get()!!.applicationContext.showToast(error)
+        activity?.applicationContext?.showToast(error)
```
You don't have to validate activity is null

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,44 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
 
     fun hideLoading() {
-        activityRef.get()!!.progressBar.visibility = View.GONE
+        activity?.progressBar?.visibility = View.GONE
```
You don't have to validate activity is null

gerardosuarezg commented on app/src/main/java/com/puzzlebench/clean_marvel_kotlin/presentation/mvp/CharacterView.kt:
```diff
@@ -7,47 +7,44 @@ import com.puzzlebench.clean_marvel_kotlin.domain.model.Character
     fun showLoading() {
-        activityRef.get()!!.progressBar.visibility = View.VISIBLE
-
+        activity?.progressBar?.visibility = View.VISIBLE
```
You don't have to validate activity is null

gerardosuarezg commented on app/src/main/res/layout/fragment_dialog_detail_character.xml:
```diff
@@ -0,0 +1,31 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent"
+        />
+    <RelativeLayout
```
Please remove this RelativeLayout. Is enough with the above one.

taidy-marrugo committed:
comments fixed

-remove wild card imports
- use $ template
-remove aditional reference of activity
-remove activity and fragment null validation
-remove relativeLayout

taidy-marrugo committed:
reverse to concat with plus
